clang/test/CheckedCRewriter/arrboth.c

```diff
@@ -124,8 +124,9 @@ z += 2;
 return z; }
 //CHECK_NOALL: int * sus(int *x, _Ptr<int> y) {
 //CHECK_NOALL:         int *z = calloc(5, sizeof(int)); 
-//CHECK_ALL: int * sus(int *x, _Ptr<int> y) {
-//CHECK_ALL:         int *z = calloc(5, sizeof(int)); 
+//CHECK_ALL: _Nt_array_ptr<int> sus(int *x, _Ptr<int> y) {
+//CHECK_ALL:         _Nt_array_ptr<int> z =  calloc(5, sizeof(int)); 
+//CHECK_ALL: _Array_ptr<int> p = z;
 
```

Comment:
I don't understand this one. It looks like `p` is declared in the `for` block. How is this rewriting happening?

Reply:
This beats me. I've no clue why it converts this way. Investigating now

Reply:
OK. It's because of how the tool converts initializations that happen in for loops, I think, but I'm not familiar with that part of the code yet so this might have more significance for you.
The following code: 
```
 int * foo() { 
    int *z = calloc(5, sizeof(int)); 
    for(int i = 0, *p = z, fac = 1; i < 5; ++i, p++, fac *= i) 
    { *p = fac; }
    return z;
}
``` 
converts to: 
```
_Nt_array_ptr<int> foo(void) { 
    _Nt_array_ptr<int> z =  calloc(5, sizeof(int)); 
    forint i = 0;
_Array_ptr<int> p = z;
int fac = 1;
( i < 5; ++i, p++, fac *= i) 
    { *p = fac; }
    return z;
}
``` 
with alltypes. It seems that the conversion tool takes all of the initialization expressions in the for loop and brings them out in order to convert them. I think it'd be fine, except it's a little weird that the `for` gets pulled out with the `int i = 0;` I think the idea was to do something like: 
```
int i = 0; 
_Array_ptr<int> p = z; 
int fac = 1; 
for(; i<5; ++i, p++, fac *= i) { ... } 
```
Should I open an issue for this?

Reply:
Yes, this is a bug. Not even valid C, much less valid Checked C. I'm hesitant for you to check in these tests that we know are (syntactically) wrong. Is it possible to change the test a little? E.g., change your test to not generate declarations in the `for` itself; move them outside, e.g., 
```
for(int i = 0, *p = z, fac = 1 ...
```
instead have it do
```
int i, *p, fac;
for(i = 0, *p = z, fac = 1; ... 
```
Then make an issue for the problem.

Reply:
Yes, that makes the most sense. I'll make an issue for this and then change up the existing tests.

Reply:
I notice that this test is a little bogus, since you are making a variable `p` that is never initialized, and then you deref it. Am I right about that? I'm wondering why/if this makes sense.

Reply:
p is set to z in the initialization portion of the for loop, and then by incrementing p, we iterate through the 5-element array and assign values to z[0] - z[4]

Reply:
I see. I was thinking that you were setting `*p` to `z` not `p` to `z`. But I see that I was just confused about the combination of initialization and a list of variables. So my suggested fix is slightly wrong. Should be
```
int i, *p, fac;
for(i = 0, p = z, fac = 1; ... 
```